### PR TITLE
Expose allocator interface to publishers and subscriptions

### PR DESCRIPTION
**Pull request by jacquelinekay**
Connects to #103 

Not ready for full review yet, but I'm opening this PR to get feedback about the approach.

An ament repo with an alternative allocator is available here:

https://github.com/ros2/tlsf_cpp

### CODE REVIEW

**Comment by wjwwood** [comment]
Looks like a file called `allocator_deleter.hpp` is missing, or is that what will be in the external package?

**Comment by jacquelinekay** [comment]
oops, yeah, thanks for catching that

File: rclcpp/include/rclcpp/publisher.hpp [hunk 1]
```diff
@@ -40,10 +40,116 @@ class Node;
...
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS_NOT_COPYABLE(PublisherBase);
+  PublisherBase(
+    std::shared_ptr<rmw_node_t> node_handle,
+    rmw_publisher_t * publisher_handle,
+    std::string topic,
+    size_t queue_size
+  ) : node_handle_(node_handle), publisher_handle_(publisher_handle),
+      topic_(topic), queue_size_(queue_size),
+    intra_process_publisher_handle_(nullptr),
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/publisher.hpp]: Inconsistent indentation in the last two lines.

File: rclcpp/include/rclcpp/allocator_deleter.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  AllocatorDeleter(const AllocatorDeleter<U, Allocator> & a){
+    allocator_ = a.allocator_;
+  }
+
+  void operator()(T* ptr) {
+    std::allocator_traits<Allocator>::destroy(*allocator_, ptr);
+    std::allocator_traits<Allocator>::deallocate(*allocator_, ptr, sizeof(T));
+  }
+
+  Allocator * allocator_;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/allocator_deleter.hpp]: Can this be private / protected?

File: rclcpp/include/rclcpp/allocator_deleter.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+#ifndef RCLCPP_RCLCPP_ALLOCATOR_DELETER_HPP_
+#define RCLCPP_RCLCPP_ALLOCATOR_DELETER_HPP_
+
+#include <memory>
+
+template<typename T, typename Allocator>
+class AllocatorDeleter
+{
+public:
+  AllocatorDeleter() : allocator_(new Allocator) {
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/allocator_deleter.hpp]: `uncrustify` should complain about the trailing `{` for the function.

Same for the other functions in this file.

File: rclcpp/include/rclcpp/allocator_deleter.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  AllocatorDeleter(Allocator* a) : allocator_(a) {
+  }
+
+  template<typename U>
+  AllocatorDeleter(const AllocatorDeleter<U, Allocator> & a){
+    allocator_ = a.allocator_;
```
Comment (dirk-thomas) [inline rclcpp/include/rclcpp/allocator_deleter.hpp]: If the allocator_ pointer is just copied into this instance won't both instances try to destroy it at the end?

**Comment by jacquelinekay** [comment]
I'm in the process of breaking up what independent parts of this PR that I can.

I may do yet another overhaul of the API because I still haven't sketched out the codepath for passing allocators to incidental STL structures used by Publisher, Subscribers, etc.

**Comment by jacquelinekay** [comment]
Closing this because I have a different approach that I think is better.

**Pull request closed without merging**